File: src/test/java/io/github/binout/soccer/infrastructure/persistence/mongo/MongoSeasonRepositoryTest.java
package io.github.binout.soccer.infrastructure.persistence.mongo;

import io.github.binout.soccer.domain.season.Season;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.mongolink.test.MongolinkRule;

import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * LECTRA
 *
 * @author b.prioux
 */
public class MongoSeasonRepositoryTest {

    @Rule
    public MongolinkRule mongolinkRule =  MongolinkRule.withPackage(getClass().getPackage().getName());

    private MongoSeasonRepository repository;

    @Before
    public void initRepository() {
        repository = new MongoSeasonRepository();
        repository.mongoSession = mongolinkRule.getCurrentSession();
    }

    @Test
    public void should_persist_season() {
        repository.add(new Season("2016"));
        repository.mongoSession.flush();

        Optional<Season> season = repository.byName("2016");
        assertThat(season).isPresent();
        assertThat(season.get().id()).isNotNull();
        assertThat(season.get().friendlyMatches().count()).isZero();
        assertThat(season.get().leagueMatches().count()).isZero();
    }

}

File: src/test/kotlin/io/github/binout/soccer/infrastructure/persistence/mongo/MongoSeasonRepositoryTest.kt
package io.github.binout.soccer.infrastructure.persistence.mongo

import io.github.binout.soccer.domain.date.FriendlyMatchDate
import io.github.binout.soccer.domain.date.LeagueMatchDate
import io.github.binout.soccer.domain.player.Player
import io.github.binout.soccer.domain.player.PlayerName
import io.github.binout.soccer.domain.season.Season
import io.github.binout.soccer.infrastructure.persistence.MongoConfiguration
import io.github.binout.soccer.infrastructure.persistence.MongoSeasonRepository
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.time.LocalDate
import java.time.Month
import java.util.*

class MongoSeasonRepositoryTest {

    private lateinit var repository: MongoSeasonRepository

    @BeforeEach
    fun initRepository() {
        repository = MongoSeasonRepository(MongoConfiguration().database())
    }

    @Test
    fun should_persist_season() {
        repository.replace(Season("2016"))

        val season = repository.byName("2016")
        assertThat(season).isNotNull
        assertThat(season!!.friendlyMatches()).isEmpty()
        assertThat(season.leagueMatches()).isEmpty()
    }

    @Test
    fun should_persist_season_with_matches() {
        val season2017 = Season("2017")

        val friendlyPlayers = players(10)
        val friendlyMatchDate = FriendlyMatchDate(LocalDate.of(2017, Month.APRIL, 12))
        friendlyPlayers.forEach { friendlyMatchDate.present(it)}
        season2017.addFriendlyMatch(friendlyMatchDate, friendlyPlayers)

        val leaguePlayers = players(7)
        val leagueMatchDate = LeagueMatchDate(LocalDate.of(2017, Month.APRIL, 12))
        leaguePlayers.forEach { leagueMatchDate.present(it) }
        season2017.addLeagueMatch(leagueMatchDate, leaguePlayers)

        repository.replace(season2017)

        val season = repository.byName("2017")
        assertThat(season).isNotNull
        assertThat(season!!.friendlyMatches()).hasSize(1)
        assertThat(season.leagueMatches()).hasSize(1)
    }

    private fun players(nb: Int) : Set<Player> = (0 until nb)
            .map { UUID.randomUUID().toString() }
            .map { PlayerName(it) }
            .map { Player(it) }.toSet()
}